server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2729,13 +2731,12 @@ void resetEngineToGlobalCheckpoint() throws IOException {
             newEngine = createNewEngine(newEngineConfig());
             active.set(true);
         }
+        newEngine.advanceMaxSeqNoOfUpdatesOrDeletes(seqNoStats.getMaxSeqNo());
```

Comment:
I think it will be easier to understand if we use `seqNoStats.getGlobalCheckpoint()` which is the upper bound for the local recovery. The rest we don't care about here, right?